oshmem/mca/scoll/ucc/scoll_ucc_dtypes.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2021 Mellanox Technologies. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+#ifndef SCOLL_UCC_DTYPES_H
+#define SCOLL_UCC_DTYPES_H
+
+#include "oshmem/op/op.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/op/op.h"
+
+#include <ucc/api/ucc.h>
+
+#define SCOLL_UCC_DT_UNSUPPORTED -1
+#define SCOLL_UCC_OP_UNSUPPORTED -1
+
+static ucc_datatype_t shmem_datatype_to_ucc_dt[OSHMEM_OP_TYPE_NUMBER + 1] = {
+    UCC_DT_INT16,                 /* OSHMEM_OP_TYPE_SHORT             0 */
+    UCC_DT_INT32,                 /* OSHMEM_OP_TYPE_INT               1 */
+    UCC_DT_INT64,                 /* OSHMEM_OP_TYPE_LONG              2 */
+    UCC_DT_INT128,                /* OSHMEM_OP_TYPE_LLONG             3 */
+    UCC_DT_INT16,                 /* OSHMEM_OP_TYPE_INT16_T           4 */
+    UCC_DT_INT32,                 /* OSHMEM_OP_TYPE_INT32_T           5 */
+    UCC_DT_INT64,                 /* OSHMEM_OP_TYPE_INT64_t           6 */
+    UCC_DT_FLOAT32,               /* OSHMEM_OP_TYPE_FLOAT             7 */
+    UCC_DT_FLOAT64,               /* OSHMEM_OP_TYPE_DOUBLE            8 */
+    SCOLL_UCC_DT_UNSUPPORTED,     /* OSHMEM_OP_TYPE_LDOUBLE           9 */
+    SCOLL_UCC_DT_UNSUPPORTED,     /* OSHMEM_OP_TYPE_FCOMPLEX         10 */
+    SCOLL_UCC_DT_UNSUPPORTED,     /* OSHMEM_OP_TYPE_DCOMPLEX         11 */
+    UCC_DT_INT16,                 /* OSHMEM_OP_TYPE_FINT2            12 */
+    UCC_DT_INT32,                 /* OSHMEM_OP_TYPE_FINT4            13 */
+    UCC_DT_INT64,                 /* OSHMEM_OP_TYPE_FINT8            14 */
+    UCC_DT_FLOAT32,               /* OSHMEM_OP_TYPE_FREAL4           15 */
+    UCC_DT_FLOAT64,               /* OSHMEM_OP_TYPE_FREAL8           16 */
+    SCOLL_UCC_DT_UNSUPPORTED,     /* OSHMEM_OP_TYPE_FREAL16          17 */
+    SCOLL_UCC_DT_UNSUPPORTED      /* OSHMEM_OP_TYPE_NUMBER           18 */
+};
+
+static inline ucc_datatype_t shmem_op_to_ucc_dtype(oshmem_op_t *op)
+{
+    ucc_datatype_t dtype;
+
+    dtype = shmem_datatype_to_ucc_dt[op->dt];
+    if (dtype == (ucc_datatype_t) SCOLL_UCC_DT_UNSUPPORTED &&
+        op->dt == OSHMEM_OP_TYPE_NUMBER) {
```

Comment:
do you need the first part of conditional here? looks like op->dt == OSHMEM_OP_TYPE_NUMBER is enough since it is always implies dtype = SCOLL_UC_DT_UNSUPPORTED.

What is OSHMEM_OP_TYPE_NUMBER ? Is it signed/unsigned?

Reply:
OSHMEM_OP_TYPE_NUMBER is the largest value of an enumeration used to keep track of the datatypes used for reductions. Should always be 18.